### PR TITLE
Fix SR-2134: emit 'rawValue' bodies for enums with a generic raw value

### PR DESCRIPTION
**Pull request by karwa**
#### Resolved bug number: ([SR-2134](https://bugs.swift.org/browse/SR-2134))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 1]
```diff
@@ -82,9 +82,16 @@ static void deriveBodyRawRepresentable_raw(AbstractFunctionDecl *toRawDecl) {
 
   Type rawTy = enumDecl->getRawType();
   assert(rawTy);
+  // If the raw type is a generic parameter, unwrap it to match the type of the literals.
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: I'm not sure this is correct. You should not depend on SubstitutedType for anything other than diagnostics...

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 2]
```diff
@@ -82,9 +82,16 @@ static void deriveBodyRawRepresentable_raw(AbstractFunctionDecl *toRawDecl) {
...
+  auto sub_rawTy = dyn_cast<SubstitutedType>(rawTy.getPointer());
+  if (sub_rawTy && sub_rawTy->getReplacementType()->isTypeParameter()) {
+	  rawTy = sub_rawTy->getOriginal();
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: It looks like you're trying to convert from an interface type to an ArchetypeType -- you should instead use ArchetypeBuilder::mapTypeIntoContext() for that

File: test/Sema/enum_raw_representable_generic.swift [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+let _ : Int64 = Foo.a.rawValue
+let _ : UInt8  = Foo.a.rawValue
+let _ : UInt16 = Foo.a.rawValue
+let _ : UInt32 = Foo.a.rawValue
+let _ : UInt64 = Foo.a.rawValue
+
+guard case .c = Foo(rawValue: UInt8(42))!  else { fatalError() }
+guard case .c = Foo(rawValue: UInt16(42))! else { fatalError() }
+guard case .c = Foo(rawValue: UInt32(42))! else { fatalError() }
+guard case .c = Foo(rawValue: UInt64(42))! else { fatalError() }
```
Comment (slavapestov) [inline test/Sema/enum_raw_representable_generic.swift]: Can you add a test that hits the FIXME case? It's not clear to me what that would mean.

Comment (karwa) [inline test/Sema/enum_raw_representable_generic.swift]: The FIXME case is for when the types of the cases don't match, e.g:
`enum Foo<T:UnsignedInteger> : T {                                                                                                                                                                                 case a = 10                                                                                                                                                                                                             case b = "hello"                                                                                                                                                                                                        }`

I'll try and add some proper diagnostics instead. I'm just not that familiar with the compiler yet...

Comment (slavapestov) [inline test/Sema/enum_raw_representable_generic.swift]: Thanks -- what happens currently if the types don't match up? You should add a test case either way.

No worries, the codebase is huge so we're all learning :)

Comment (karwa) [inline test/Sema/enum_raw_representable_generic.swift]: @slavapestov updated as you commented.

When the FIXME case happens, I assume it'll be as before - we'll emit a declaration but no definition, so it'll fail at link time.

Comment (slavapestov) [inline test/Sema/enum_raw_representable_generic.swift]: What happens if the declared raw type is not generic, but the raw types of the elements don't match up? Do we also crash?

If not, then the problem is that we're being too lax when type checking the *elements*.

Comment (karwa) [inline test/Sema/enum_raw_representable_generic.swift]: Yes, this also causes a crash:
```
enum Foo : UInt8 { case a = 10, case b = "hello" }
```

Comment (slavapestov) [inline test/Sema/enum_raw_representable_generic.swift]: Nice. It looks like the crash occurs because its trying to type check a function with no body, which is not allowed.

Can we push the check for 'all elements have the same raw type' up a level, so that we don't attempt to derive the rawRepresentable conformance at all in that case?

Similar to how we don't derive an Equatable conformance for enums that have payload cases (we could, eventually, but the point is we have control over when to derive the conformance).

**Comment by karwa** [comment]
@slavapestov Okay, I've improved the RawRepresentable derived conformance checking. I've tried it with badly-typed cases (generic and non-generic) and it doesn't crash any more.

I haven't run the actual unit tests though

**Comment by tkremenek** [comment]
@swift-ci test

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 4]
```diff
@@ -357,16 +353,15 @@ ValueDecl *DerivedConformance::deriveRawRepresentable(TypeChecker &tc,
                                                       Decl *parentDecl,
                                                       NominalTypeDecl *type,
                                                       ValueDecl *requirement) {
-  // Check preconditions. These should already have been diagnosed by
-  // type-checking but we may still get here after recovery.
+  // Check preconditions for RawRepresentable conformance.
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: It would be great to put the precondition check in its own function, since the same logic is duplicated in the two overloads of DerivedConformance::deriveRawRepresentable()

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 5]
```diff
@@ -376,8 +371,13 @@ ValueDecl *DerivedConformance::deriveRawRepresentable(TypeChecker &tc,
   if (enumDecl->getAllElements().empty())
     return nullptr;
 
-  for (auto elt : enumDecl->getAllElements())
+  // The enum elements must all have the same type (and must match the enum type).
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: I think you're missing the actual type check here?

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 6]
```diff
@@ -376,8 +371,13 @@ ValueDecl *DerivedConformance::deriveRawRepresentable(TypeChecker &tc,
...
+  for (auto elt : enumDecl->getAllElements()) {
     tc.validateDecl(elt);
+	if (elt->isInvalid()) {
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: Indentation

**Comment by slavapestov** [comment]
@karwa Thanks! So it looks like validateDecl() checks that all the types match for us -- can you just add a comment to that effect (and also, see my note about the duplicated checks).

Other than that, this looks good, thanks!

**Comment by slavapestov** [comment]
@karwa One more thing, can you also add tests for the invalid cases that you fixed?

**Comment by tkremenek** [comment]
@swift-ci test os x

**Comment by tkremenek** [comment]
@swift-ci test linux

**Comment by karwa** [comment]
There is one test that still isn't working - decl/protocol/conforms/placement.swift

In placement_2.swift:
```
enum MFSynthesizedEnum1 : Int { case a }
```

In placement.swift:
```
extension MFSynthesizedEnum1 : RawRepresentable { }
```

The comments say something about 'suppressing' synthesized conformances, which I don't understand. What exactly is supposed to happen in this case? An error?

**Comment by slavapestov** [comment]
Interesting. If I create a file `a.swift` containing `enum MFSynthesizedEnum1 : Int { case a }`, and a file `main.swift` containing `extension MFSynthesizedEnum1 : RawRepresentable { }`, then `swiftc a.swift main.swift` crashes.

I wonder why the test passes, then.

In any case, the extension should be a no-op, or perhaps the members should be synthesized in the extension and not in the original type.

**Comment by karwa** [comment]
I have come across so many tests which inexplicably haven't been failing before, I can't understand it.

The mismatched enum types? That's already in test/Parse/enum.swift -- but try it in the interpreter, and it crashes.
Generic raw values? That's already in test/Generics/inheritance.swift -- but it wasn't working, hence the bug report SR-2134

**Comment by slavapestov** [comment]
> The mismatched enum types? That's already in test/Parse/enum.swift -- but try it in the interpreter, and it crashes.

I don't see a test here that exercises that path. Maybe I'm missing something, but it could also be that the *combination* of a generic raw value and mismatched types crashes.

> Generic raw values? That's already in test/Generics/inheritance.swift -- but it wasn't working, hence the bug report SR-2134

I think the difference between this test and yours is that you're giving the cases explicit raw values using '='. Try playing with that?

**Comment by slavapestov** [comment]
Thanks for digging into this regardless -- figuring out these corner cases and adding detailed tests for them is really, really important for the long-term health of the language.

File: test/Parse/enum.swift [hunk 7]
```diff
@@ -308,12 +308,12 @@ enum NonliteralRawValue : Int {
...
-enum RawTypeWithPayload : Int { // expected-note {{declared raw type 'Int' here}} expected-note {{declared raw type 'Int' here}}
+enum RawTypeWithPayload : Int { // expected-error {{type 'RawTypeWithPayload' does not conform to protocol 'RawRepresentable'}} expected-note {{declared raw type 'Int' here}} expected-note {{declared raw type 'Int' here}}
   case Powell(Int) // expected-error {{enum with raw type cannot have cases with arguments}}
   case Terwilliger(Int) = 17 // expected-error {{enum with raw type cannot have cases with arguments}}
 }
 
-enum RawTypeMismatch : Int {
+enum RawTypeMismatch : Int { // expected-error {{type 'RawTypeMismatch' does not conform to protocol 'RawRepresentable'}}
   case Barbur = "foo" // expected-error {{}}
 }
```
Comment (karwa) [inline test/Parse/enum.swift]: Actually, it's this one.

Comment (slavapestov) [inline test/Parse/enum.swift]: It could be that in the interpreter, we go ahead and type check the bodies all synthesized members, but in the test, we do not.

Remember that bodies of synthesized members are built lazily, and we only crashed because we'd hit a case where a synthesized member subsequently had no body.

**Comment by slavapestov** [comment]
If you can think of a good way to re-organize the test, go ahead :)

Don't kill yourself trying to understand some of the emergent behavior that comes up -- it's more important to make the existing tests pass, and add new tests for the problems you found.

**Comment by slavapestov** [comment]
Do you mind squashing the commits into one patch, and referencing the bug number from the commit message too?

**Comment by karwa** [comment]
Yes, I will, once I've fixed `decl/protocol/conforms/placement.swift`

**Comment by karwa** [comment]
Okay, so it's been a day of learning :)

The problem is that if the extension is in the primary file and the enum itself in some other file, we'll try to synthesize RawRep before the enum has been type-checked, before implicit cases have been assigned values, etc. The other way around (enum in primary file) works.

I'm going to try and fix it by removing the suppression behaviour (even though it seems intentional). So declaring RawRepresentable will be a no-op, although of course you can provide your own implementation if you don't want the synthesised one. It doesn't really make sense that:
```
enum Foo : Int { case a }
```
becomes:
```
enum Foo : Int, RawRepresentable {
    case a
    Typealias RawValue = Int
    init?(rawValue: Int)
    var rawValue: Int { get }
    var hashValue: Int { get }
}
```

Meanwhile,

```
enum Foo : Int { case a = 10 }
extension Foo : RawRepresentable {}
```
becomes:
```
enum Foo : Int {
    case a
    var hashValue: Int { get }
}
```

Feels like a hack for some standard-library thing. @gribozavr, any idea?

**Comment by karwa** [comment]
After a little rethink, I settled on the idea that declaring a raw type (or giving a value to an element, which already requires that you explicitly specify a raw type in the inheritance clause) should count as an explicit declaration to synthesise RawRepresentable conformance.

The inheritance clause and label statements in enums are really just syntactic sugar. It doesn't change anything about the underlying storage of the enum -- and even if it did, you wouldn't be able to change that in an extension. That would be like re-parenting a class.

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 8]
```diff
@@ -82,11 +82,11 @@ static void deriveBodyRawRepresentable_raw(AbstractFunctionDecl *toRawDecl) {
...
   Type rawTy = enumDecl->getRawType();
   assert(rawTy);
+  rawTy = ArchetypeBuilder::mapTypeIntoContext(toRawDecl, rawTy);
+
   for (auto elt : enumDecl->getAllElements()) {
-    if (!elt->getTypeCheckedRawValueExpr() ||
-        !elt->getTypeCheckedRawValueExpr()->getType()->isEqual(rawTy)) {
-      return;
-    }
+    assert(elt->getTypeCheckedRawValueExpr() && "Enum element has no literal - missing a call to checkEnumRawValues()");
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: Lines should be shorter than 80 characters

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 9]
```diff
@@ -257,18 +255,10 @@ static ConstructorDecl *deriveRawRepresentable_init(TypeChecker &tc,
...
                                        KnownProtocolKind::Equatable);
-  if (!equatableProto)
-    return nullptr;
-
-  if (!tc.conformsToProtocol(rawType, equatableProto, enumDecl, None)) {
-    SourceLoc loc = enumDecl->getInherited()[0].getSourceRange().Start;
-    tc.diagnose(loc, diag::enum_raw_type_not_equatable, rawType);
-    return nullptr;
-  }
+  assert(equatableProto);
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: This will emit an unused variable warning if asserts are off.

Add a `(void) equatableProto;` right after to silence this (and confirm by compiling with `./build-script -r`)

**Comment by slavapestov** [comment]
This latest version looks good. Please just fix the warnings and line length and we're good to go!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/AST/ArchetypeBuilder.cpp [hunk 10]
```diff
@@ -2028,7 +2028,10 @@ Type ArchetypeBuilder::mapTypeIntoContext(Module *M,
   if (!canType->hasTypeParameter())
     return type;
 
-  assert(genericParams && "dependent type in non-generic context");
+  if (!genericParams) {
+    // Failed: dependent type in non-generic context.
+    return Type();
+  }
```
Comment (karwa) [inline lib/AST/ArchetypeBuilder.cpp]: I'm not sure if this is the right thing to do, but the alternative is to duplicate these checks before attempting to map the type. There's an argument that this function shouldn't assert on failure conditions at all - in most other cases, the type-checker keeps going even after discovering invalid declarations.

Comment (slavapestov) [inline lib/AST/ArchetypeBuilder.cpp]: I added these checks recently. There is a distinction between failure conditions that arise from invalid user input and failure conditions that arise from bugs inside the type checker. The latter should assert.

Note that if there are no generic parameters, we should not hit the assert, because the given type should not contain any type parameters. So if canType->hasTypeParameter() is true but genericParams is null, we have a bug.

Are you seeing this in one of the tests?

File: lib/Sema/DerivedConformanceRawRepresentable.cpp [hunk 11]
```diff
@@ -259,6 +261,7 @@ static ConstructorDecl *deriveRawRepresentable_init(TypeChecker &tc,
                                        KnownProtocolKind::Equatable);
   assert(equatableProto);
   assert(tc.conformsToProtocol(rawType, equatableProto, enumDecl, None));
+  (void)equatableProto; // silence unused variable warning when assertions off.
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceRawRepresentable.cpp]: The comment is not necessary, this is a common idiom in our code. :)

File: lib/AST/ArchetypeBuilder.cpp [hunk 12]
```diff
@@ -2087,7 +2090,10 @@ Type ArchetypeBuilder::mapTypeOutOfContext(ModuleDecl *M,
   if (!canType->hasArchetype())
     return type;
 
-  assert(genericParams && "dependent type in non-generic context");
+  if (!genericParams) {
+    // Failed: dependent type in non-generic context.
+    return Type();
```
Comment (slavapestov) [inline lib/AST/ArchetypeBuilder.cpp]: Yeah, so returning a null type isn't really any better than asserting -- you're just pushing the crash downstream, where it is harder to figure out the root case.

Which crasher regressed and what was the error?

File: lib/AST/ArchetypeBuilder.cpp (hunk 10 above)
Comment (karwa) [inline lib/AST/ArchetypeBuilder.cpp]: 09650-swift-typebase-getcanonicaltype.swift

From the fuzzer, so it's not valid code anyway.

Comment (slavapestov) [inline lib/AST/ArchetypeBuilder.cpp]: The problem then is we have an enum with a raw type that's generic, but no generic parameter list.

This means some other part of the compiler upstream of you constructed an invalid declaration. If you figure out what's going on and fix it there, you might be able to squash a few more compiler crashers.

Or, you can just XFAIL the compiler crasher for now. Having one or two regressions isn't the end of the world, especially since you improved matters in other respects.

**Comment by slavapestov** [comment]
@karwa Are you still working on this? Let me know if you need some help!

**Comment by karwa** [comment]
@slavapestov yes! Been ill, will fix it up tonight

**Comment by slavapestov** [comment]
No worries, get well soon!

**Comment by karwa** [comment]
@slavapestov Fixed, and while I was there I didn't like the look of the new errors (not very descriptive where the RawRepresentable conformance comes from), so I made them more descriptive.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test os x

File: lib/Sema/TypeCheckProtocol.cpp [hunk 13]
```diff
@@ -4718,6 +4730,19 @@ void TypeChecker::checkConformancesInContext(DeclContext *dc,
              dc->getDeclaredTypeInContext(),
              diag.Protocol->getName());
 
+	// Special case: explain that 'RawRepresentable' conformance
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Indentation

File: validation-test/compiler_crashers/09650-swift-typebase-getcanonicaltype.swift [hunk 14]
```diff
@@ -5,7 +5,7 @@
 // See http://swift.org/LICENSE.txt for license information
 // See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
-// RUN: not %target-swift-frontend %s -parse
+// RUN: not --crash %target-swift-frontend %s -parse
```
Comment (slavapestov) [inline validation-test/compiler_crashers/09650-swift-typebase-getcanonicaltype.swift]: I'll see about fixing this crash

**Comment by slavapestov** [comment]
Thanks!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS and merge

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
Alright @karwa, I think you can close that JIRA now :)
